lib/internal/streams/pipeline.js

```diff
@@ -158,10 +162,11 @@ function pipeline(...streams) {
     }
   }
 
-  function wrap(stream, reading, writing, final) {
-    destroys.push(destroyer(stream, reading, writing, (err) => {
-      finish(err, final);
-    }));
+  function wrap(stream, reading, writing) {
+    const fn = destroyer(stream, reading, writing, (err) => {
+      finish(err);
+    });
+    destroys.push(fn);
```

Comment:
What about:
```js
 function wrap(stream, reading, writing, final) {
    finishCount++;
    destroys.push(destroyer(stream, reading, writing, (err) => {
      finishCount--;
      finish(err, finishCount === 0);
    }));
 }
```

Reply:
I've not been able to make it work. There are a few edge case, as every call to `pump()` will call `finish()` in case of an error, and this is exactly the call to `finish()` we need to wait for.

Reply:
How about something like this? https://github.com/nxtedition/node/pull/15/files

Reply:
In your code, `finish()` is always called in a `finally` block. This have all the `.destroy()` methods in streams called, potentially allocating a lot of Errors objects (and maybe erroring). I would like to avoid this.

Reply:
@mcollina: updated, PTAL